File: player/windows/src/PetsGraph.Core/PassiveBehaviorSession.cs
using System.Security.Cryptography;

namespace PetsGraph.Core;

public sealed record PetPlaybackPresentation(
    string ClipId,
    int FrameIndex,
    string CurrentNodeId,
    bool IsTransition,
    IReadOnlyList<string> PreloadClipIds);

public sealed class PassiveBehaviorSession
{
    private readonly LoadedPetPack package;
    private readonly IReadOnlyDictionary<string, PetGraphNode> nodes;
    private readonly IReadOnlyDictionary<string, PetGraphEdge[]> outgoing;
    private readonly PetGraphNode[] eligibleNodes;
    private SplitMix64 random;
    private string currentNodeId;
    private string currentClipId;
    private double segmentStartedAt;
    private double dwellDeadline;
    private readonly List<PetGraphEdge> pendingEdges = [];
    private double? scheduledExitAt;
    private bool hiddenRequested;
    private bool paused;
    private int frozenFrameIndex;

    public PassiveBehaviorSession(LoadedPetPack package, double startedAt, ulong? seed = null)
    {
        this.package = package;
        nodes = package.Graph.Nodes.ToDictionary(static node => node.Id, StringComparer.Ordinal);
        outgoing = package.Graph.Edges.GroupBy(static edge => edge.From)
            .ToDictionary(static group => group.Key,
                static group => group.OrderBy(static edge => edge.Id, StringComparer.Ordinal).ToArray(),
                StringComparer.Ordinal);
        eligibleNodes = package.Graph.Nodes.Where(static node => node.AutonomousEligible)
            .OrderBy(static node => node.Id, StringComparer.Ordinal).ToArray();
        currentNodeId = package.Manifest.Stage.DefaultNode;
        currentClipId = nodes[currentNodeId].LoopClip
            ?? throw Invalid("invalid_graph", "default node has no loop clip");
        segmentStartedAt = startedAt;
        var randomSeed = seed ?? SeedFromDigest(package.ArchiveSha256) ^ RandomSeed();
        random = new(randomSeed);
        dwellDeadline = startedAt + RandomDwell(currentNodeId);
    }

    public bool IsPaused => paused;
    public bool IsTransitioning => package.Clips[currentClipId].Type == "transition";
    public bool ShouldTickWhenHidden => hiddenRequested && !paused;
    public string CurrentStableNodeId => currentNodeId;

    public void SetVisible(bool visible, double now)
    {
        if (visible)
        {
            hiddenRequested = false;
            if (paused)
            {
                paused = false;
                segmentStartedAt = now;
                scheduledExitAt = null;
                pendingEdges.Clear();
                frozenFrameIndex = 0;
                dwellDeadline = now + RandomDwell(currentNodeId);
            }
            return;
        }
        _ = Update(now);
        hiddenRequested = true;
        if (!IsTransitioning)
        {
            PauseAtStableNode(now);
        }
    }

    public PetPlaybackPresentation Update(double now)
    {
        if (!double.IsFinite(now))
        {
            throw Invalid("invalid_time", "runtime time is not finite");
        }
        if (paused)
        {
            return Presentation(frozenFrameIndex);
        }
        AdvanceCompletedSegments(now);
        if (!IsTransitioning && !hiddenRequested)
        {
            if (pendingEdges.Count == 0 && now >= dwellDeadline)
            {
                PlanNextTarget(now);
            }
            if (scheduledExitAt is { } exitAt && now >= exitAt && pendingEdges.Count != 0)
            {
                var edge = pendingEdges[0];
                pendingEdges.RemoveAt(0);
                scheduledExitAt = null;
                currentClipId = edge.Clip;
                segmentStartedAt = exitAt;
                AdvanceCompletedSegments(now);
            }
        }
        var clip = package.Clips[currentClipId];
        var elapsed = Math.Max(0, now - segmentStartedAt);
        var rawFrame = (int)Math.Floor(elapsed * clip.FrameRate.FramesPerSecond);
        var frameIndex = clip.Type == "loop" ? rawFrame % clip.FrameCount : Math.Min(rawFrame, clip.FrameCount - 1);
        return Presentation(frameIndex);
    }

    private void AdvanceCompletedSegments(double now)
    {
        var safety = 0;
        while (IsTransitioning)
        {
            if (++safety > package.Graph.Edges.Length + 1)
            {
                throw Invalid("behavior_cycle", "transition advancement exceeded the graph budget");
            }
            var clip = package.Clips[currentClipId];
            var completedAt = segmentStartedAt + clip.DurationSeconds;
            if (now < completedAt)
            {
                return;
            }
            var edge = package.Graph.Edges.Single(item => item.Clip == clip.Id);
            currentNodeId = edge.To;
            var node = nodes[currentNodeId];
            if (node.Role == "gateway")
            {
                if (pendingEdges.Count == 0 || pendingEdges[0].From != node.Id)
                {
                    throw Invalid("invalid_graph", "transition path stopped at a gateway");
                }
                currentClipId = pendingEdges[0].Clip;
                pendingEdges.RemoveAt(0);
                segmentStartedAt = completedAt;
                continue;
            }
            currentClipId = node.LoopClip ?? throw Invalid("invalid_graph", "dwell node has no loop");
            segmentStartedAt = completedAt;
            if (hiddenRequested)
            {
                pendingEdges.Clear();
                scheduledExitAt = null;
                PauseAtStableNode(completedAt);
                return;
            }
            if (pendingEdges.Count != 0)
            {
                if (pendingEdges[0].From != node.Id)
                {
                    throw Invalid("invalid_graph", "planned path is discontinuous");
                }
                scheduledExitAt = NextSafeExit(completedAt, completedAt);
            }
            else
            {
                dwellDeadline = completedAt + RandomDwell(currentNodeId);
            }
            return;
        }
    }

    private void PlanNextTarget(double now)
    {
        if (eligibleNodes.Length <= 1)
        {
            dwellDeadline = now + RandomDwell(currentNodeId);
            return;
        }
        var candidates = eligibleNodes.Where(node =>
            !package.Behavior.Timing.AvoidImmediateRepeat || node.Id != currentNodeId).ToArray();
        var target = WeightedTarget(candidates);
        var path = ShortestPath(currentNodeId, target.Id);
        if (path.Count == 0)
        {
            throw Invalid("invalid_graph", "autonomous target has no transition path");
        }
        pendingEdges.AddRange(path);
        scheduledExitAt = NextSafeExit(segmentStartedAt, now);
    }

    private PetGraphNode WeightedTarget(IReadOnlyList<PetGraphNode> candidates)
    {
        var weighted = candidates.Select(node => (Node: node,
            Weight: package.Behavior.NodeWeights.GetValueOrDefault(node.Id, 1) *
                    package.Behavior.SceneWeights.GetValueOrDefault(node.Scene, 1))).ToArray();
        var total = weighted.Sum(static item => item.Weight);
        if (!double.IsFinite(total) || total <= 0)
        {
            throw Invalid("invalid_behavior", "autonomous target weights are invalid");
        }
        var selection = random.NextUnit() * total;
        foreach (var item in weighted)
        {
            selection -= item.Weight;
            if (selection < 0)
            {
                return item.Node;
            }
        }
        return weighted[^1].Node;
    }

    private IReadOnlyList<PetGraphEdge> ShortestPath(string source, string target)
    {
        var queue = new Queue<string>();
        queue.Enqueue(source);
        var visited = new HashSet<string>(StringComparer.Ordinal) { source };
        var predecessor = new Dictionary<string, PetGraphEdge>(StringComparer.Ordinal);
        while (queue.TryDequeue(out var current))
        {
            foreach (var edge in outgoing.GetValueOrDefault(current, []))
            {
                if (!visited.Add(edge.To))
                {
                    continue;
                }
                predecessor[edge.To] = edge;
                if (edge.To == target)
                {
                    var result = new List<PetGraphEdge>();
                    var cursor = target;
                    while (cursor != source)
                    {
                        var prior = predecessor[cursor];
                        result.Add(prior);
                        cursor = prior.From;
                    }
                    result.Reverse();
                    return result;
                }
                queue.Enqueue(edge.To);
            }
        }
        return [];
    }

    private double NextSafeExit(double loopStartedAt, double now)
    {
        var clip = package.Clips[currentClipId];
        if (clip.Type != "loop" || clip.SafeExitFrames.Length == 0)
        {
            throw Invalid("invalid_safe_exit", "current dwell loop has no safe exit");
        }
        var frameDuration = clip.FrameRate.Denominator / (double)clip.FrameRate.Numerator;
        var cycleDuration = clip.FrameCount * frameDuration;
        var elapsed = Math.Max(0, now - loopStartedAt);
        var firstCycle = Math.Max(0, (int)Math.Floor(elapsed / cycleDuration));
        for (var cycle = firstCycle; cycle <= firstCycle + 1; cycle++)
        {
            foreach (var frame in clip.SafeExitFrames)
            {
                var boundary = loopStartedAt + cycle * cycleDuration + (frame + 1) * frameDuration;
                if (boundary >= now - 0.0000001)
                {
                    return boundary;
                }
            }
        }
        throw Invalid("invalid_safe_exit", "could not schedule a safe loop exit");
    }

    private void PauseAtStableNode(double now)
    {
        paused = true;
        pendingEdges.Clear();
        scheduledExitAt = null;
        var clip = package.Clips[currentClipId];
        var elapsed = Math.Max(0, now - segmentStartedAt);
        frozenFrameIndex = (int)Math.Floor(elapsed * clip.FrameRate.FramesPerSecond) % clip.FrameCount;
    }

    private PetPlaybackPresentation Presentation(int frameIndex)
    {
        var preload = new HashSet<string>(StringComparer.Ordinal);
        if (pendingEdges.Count != 0)
        {
            preload.Add(pendingEdges[0].Clip);
            var finalNode = nodes[pendingEdges[^1].To];
            if (finalNode.LoopClip is { } loop)
            {
                preload.Add(loop);
            }
        }
        return new(currentClipId, frameIndex, currentNodeId, IsTransitioning,
            preload.Order(StringComparer.Ordinal).ToArray());
    }

    private double RandomDwell(string nodeId)
    {
        if (!package.Behavior.Timing.DwellRangesSeconds.TryGetValue(nodeId, out var range) ||
            range is not [var minimum, var maximum])
        {
            throw Invalid("invalid_behavior", "current node has no dwell range");
        }
        return minimum + (maximum - minimum) * random.NextUnit();
    }

    private static ulong RandomSeed()
    {
        Span<byte> bytes = stackalloc byte[8];
        RandomNumberGenerator.Fill(bytes);
        return BitConverter.ToUInt64(bytes);
    }

    private static ulong SeedFromDigest(string digest) =>
        ulong.TryParse(digest.AsSpan(0, 16), System.Globalization.NumberStyles.HexNumber,
            System.Globalization.CultureInfo.InvariantCulture, out var seed)
            ? seed
            : 0x5045545347524150;

    private static PetPackException Invalid(string code, string detail) => new(code, detail);

    private struct SplitMix64(ulong seed)
    {
        private ulong state = seed;

        public ulong Next()
        {
            state += 0x9e3779b97f4a7c15;
            var value = state;
            value = (value ^ (value >> 30)) * 0xbf58476d1ce4e5b9;
            value = (value ^ (value >> 27)) * 0x94d049bb133111eb;
            return value ^ (value >> 31);
        }

        public double NextUnit() => (Next() >> 11) / (double)(1UL << 53);
    }
}

File: player/macos/Sources/PetsGraphV1Core/PassiveBehaviorSession.swift
import Foundation

public struct PetPlaybackPresentation: Equatable, Sendable {
  public let clipID: String
  public let frameIndex: Int
  public let currentNodeID: String
  public let isTransition: Bool
  public let preloadClipIDs: [String]
}

public final class PassiveBehaviorSession: @unchecked Sendable {
  private let package: LoadedPetPack
  private let nodes: [String: PetGraphNode]
  private let outgoing: [String: [PetGraphEdge]]
  private let eligibleNodes: [PetGraphNode]
  private var random: SplitMix64

  private var currentNodeID: String
  private var currentClipID: String
  private var segmentStartedAt: TimeInterval
  private var dwellDeadline: TimeInterval
  private var pendingEdges: [PetGraphEdge] = []
  private var scheduledExitAt: TimeInterval?
  private var hiddenRequested = false
  private var paused = false
  private var frozenFrameIndex = 0

  public init(package: LoadedPetPack, startedAt: TimeInterval, seed: UInt64? = nil) throws {
    self.package = package
    nodes = Dictionary(uniqueKeysWithValues: package.graph.nodes.map { ($0.id, $0) })
    outgoing = Dictionary(grouping: package.graph.edges, by: \.from).mapValues {
      $0.sorted { $0.id < $1.id }
    }
    eligibleNodes = package.graph.nodes.filter(\.autonomousEligible).sorted { $0.id < $1.id }
    currentNodeID = package.manifest.stage.defaultNode
    guard let initialLoop = nodes[currentNodeID]?.loopClip else {
      throw PetPackError("invalid_graph", "default node has no loop clip")
    }
    currentClipID = initialLoop
    segmentStartedAt = startedAt
    random = SplitMix64(
      seed: seed
        ?? (Self.seed(from: package.archiveSHA256) ^ UInt64.random(in: UInt64.min...UInt64.max))
    )
    dwellDeadline = startedAt
    dwellDeadline =
      try startedAt
      + Self.randomDwell(
        behavior: package.behavior,
        nodeID: currentNodeID,
        random: &random
      )
  }

  public var isPaused: Bool { paused }
  public var isTransitioning: Bool { package.clips[currentClipID]?.type == "transition" }
  public var shouldTickWhenHidden: Bool { hiddenRequested && !paused }
  public var currentStableNodeID: String { currentNodeID }

  public func cancelPlannedTransition(at now: TimeInterval) throws {
    guard now.isFinite, !isTransitioning else {
      throw PetPackError(
        "invalid_transition_state", "only a planned dwell transition can be cancelled")
    }
    pendingEdges.removeAll()
    scheduledExitAt = nil
    dwellDeadline =
      try now
      + Self.randomDwell(
        behavior: package.behavior,
        nodeID: currentNodeID,
        random: &random
      )
  }

  public func setVisible(_ visible: Bool, at now: TimeInterval) throws {
    if visible {
      hiddenRequested = false
      if paused {
        paused = false
        segmentStartedAt = now
        scheduledExitAt = nil
        pendingEdges.removeAll()
        frozenFrameIndex = 0
        dwellDeadline =
          try now
          + Self.randomDwell(
            behavior: package.behavior,
            nodeID: currentNodeID,
            random: &random
          )
      }
      return
    }

    _ = try update(at: now)
    hiddenRequested = true
    if !isTransitioning {
      pauseAtStableNode(now: now)
    }
  }

  public func update(at now: TimeInterval) throws -> PetPlaybackPresentation {
    guard now.isFinite else { throw PetPackError("invalid_time", "runtime time is not finite") }
    if paused {
      return presentation(frameIndex: frozenFrameIndex)
    }

    try advanceCompletedSegments(now: now)
    if !isTransitioning, !hiddenRequested {
      if pendingEdges.isEmpty, now >= dwellDeadline {
        try planNextTarget(now: now)
      }
      if let exitAt = scheduledExitAt, now >= exitAt, let edge = pendingEdges.first {
        pendingEdges.removeFirst()
        scheduledExitAt = nil
        currentClipID = edge.clip
        segmentStartedAt = exitAt
        try advanceCompletedSegments(now: now)
      }
    }

    let clip = try currentClip()
    let elapsed = max(0, now - segmentStartedAt)
    return presentation(frameIndex: Self.frameIndex(for: clip, elapsed: elapsed))
  }

  private func advanceCompletedSegments(now: TimeInterval) throws {
    var safety = 0
    while isTransitioning {
      safety += 1
      guard safety <= package.graph.edges.count + 1 else {
        throw PetPackError("behavior_cycle", "transition advancement exceeded the graph budget")
      }
      let clip = try currentClip()
      let completedAt = segmentStartedAt + clip.durationSeconds
      guard now >= completedAt else { return }
      guard let edge = package.graph.edges.first(where: { $0.clip == clip.id }) else {
        throw PetPackError("invalid_graph", "transition clip has no graph edge")
      }
      currentNodeID = edge.to
      guard let node = nodes[currentNodeID] else {
        throw PetPackError("invalid_graph", "transition reached an unknown node")
      }
      if node.role == "gateway" {
        guard let next = pendingEdges.first, next.from == node.id else {
          throw PetPackError("invalid_graph", "transition path stopped at a gateway")
        }
        pendingEdges.removeFirst()
        currentClipID = next.clip
        segmentStartedAt = completedAt
        continue
      }
      guard let loop = node.loopClip else {
        throw PetPackError("invalid_graph", "dwell node has no loop")
      }
      currentClipID = loop
      segmentStartedAt = completedAt
      if hiddenRequested {
        pendingEdges.removeAll()
        scheduledExitAt = nil
        pauseAtStableNode(now: completedAt)
        return
      }
      if let next = pendingEdges.first {
        guard next.from == node.id else {
          throw PetPackError("invalid_graph", "planned path is discontinuous")
        }
        scheduledExitAt = try nextSafeExit(after: completedAt, now: completedAt)
      } else {
        dwellDeadline =
          try completedAt
          + Self.randomDwell(
            behavior: package.behavior,
            nodeID: currentNodeID,
            random: &random
          )
      }
      return
    }
  }

  private func planNextTarget(now: TimeInterval) throws {
    guard eligibleNodes.count > 1 else {
      dwellDeadline =
        try now
        + Self.randomDwell(
          behavior: package.behavior,
          nodeID: currentNodeID,
          random: &random
        )
      return
    }
    let candidates = eligibleNodes.filter {
      !package.behavior.timing.avoidImmediateRepeat || $0.id != currentNodeID
    }
    let target = try weightedTarget(from: candidates)
    let path = try shortestPath(from: currentNodeID, to: target.id)
    if path.isEmpty {
      dwellDeadline =
        try now
        + Self.randomDwell(
          behavior: package.behavior,
          nodeID: currentNodeID,
          random: &random
        )
      return
    }
    pendingEdges = path
    scheduledExitAt = try nextSafeExit(after: segmentStartedAt, now: now)
  }

  private func weightedTarget(from candidates: [PetGraphNode]) throws -> PetGraphNode {
    let logarithmic = candidates.map { node -> (PetGraphNode, Double) in
      let nodeWeight = package.behavior.nodeWeights[node.id] ?? 1
      let sceneWeight = package.behavior.sceneWeights[node.scene] ?? 1
      return (node, log(nodeWeight) + log(sceneWeight))
    }
    guard let maximum = logarithmic.map(\.1).max(), maximum.isFinite else {
      throw PetPackError("invalid_behavior", "autonomous target weights are invalid")
    }
    let weighted = logarithmic.map { ($0.0, exp($0.1 - maximum)) }
    let total = weighted.reduce(0) { $0 + $1.1 }
    guard total.isFinite, total > 0 else {
      throw PetPackError("invalid_behavior", "autonomous target weights are invalid")
    }
    var selection = random.nextUnit() * total
    for item in weighted {
      selection -= item.1
      if selection < 0 { return item.0 }
    }
    return weighted.last!.0
  }

  private func shortestPath(from source: String, to target: String) throws -> [PetGraphEdge] {
    var frontier = [source]
    var visited: Set<String> = [source]
    var predecessor: [String: PetGraphEdge] = [:]
    while !frontier.isEmpty {
      let current = frontier.removeFirst()
      for edge in outgoing[current, default: []] where visited.insert(edge.to).inserted {
        predecessor[edge.to] = edge
        if edge.to == target {
          var result: [PetGraphEdge] = []
          var cursor = target
          while cursor != source {
            guard let prior = predecessor[cursor] else {
              throw PetPackError("invalid_graph", "cannot reconstruct autonomous path")
            }
            result.append(prior)
            cursor = prior.from
          }
          return result.reversed()
        }
        frontier.append(edge.to)
      }
    }
    return []
  }

  private func nextSafeExit(after loopStartedAt: TimeInterval, now: TimeInterval) throws
    -> TimeInterval
  {
    let clip = try currentClip()
    guard clip.type == "loop", !clip.safeExitFrames.isEmpty else {
      throw PetPackError("invalid_safe_exit", "current dwell loop has no safe exit")
    }
    let frameDuration = Double(clip.frameRate.denominator) / Double(clip.frameRate.numerator)
    let cycleDuration = Double(clip.frameCount) * frameDuration
    let elapsed = max(0, now - loopStartedAt)
    let firstCycle = max(0, Int(floor(elapsed / cycleDuration)))
    for cycle in firstCycle...(firstCycle + 1) {
      for frame in clip.safeExitFrames {
        let boundary =
          loopStartedAt + Double(cycle) * cycleDuration
          + Double(frame + 1) * frameDuration
        if boundary > now + 0.000_000_1 { return boundary }
      }
    }
    throw PetPackError("invalid_safe_exit", "could not schedule a safe loop exit")
  }

  private func pauseAtStableNode(now: TimeInterval) {
    paused = true
    pendingEdges.removeAll()
    scheduledExitAt = nil
    let clip = package.clips[currentClipID]
    let elapsed = max(0, now - segmentStartedAt)
    frozenFrameIndex = clip.map { Self.frameIndex(for: $0, elapsed: elapsed) } ?? 0
  }

  private func currentClip() throws -> PetClip {
    guard let clip = package.clips[currentClipID] else {
      throw PetPackError("missing_clip", "behavior references a missing clip")
    }
    return clip
  }

  private func presentation(frameIndex: Int) -> PetPlaybackPresentation {
    var preload: [String] = []
    preload.append(contentsOf: pendingEdges.map(\.clip))
    if let last = pendingEdges.last, let loop = nodes[last.to]?.loopClip { preload.append(loop) }
    return PetPlaybackPresentation(
      clipID: currentClipID,
      frameIndex: frameIndex,
      currentNodeID: currentNodeID,
      isTransition: isTransitioning,
      preloadClipIDs: Array(Set(preload)).sorted()
    )
  }

  private static func randomDwell(
    behavior: PetBehavior,
    nodeID: String,
    random: inout SplitMix64
  ) throws -> TimeInterval {
    guard let range = behavior.timing.dwellRangesSeconds[nodeID], range.count == 2 else {
      throw PetPackError("invalid_behavior", "current node has no dwell range")
    }
    return range[0] + (range[1] - range[0]) * random.nextUnit()
  }

  private static func frameIndex(for clip: PetClip, elapsed: TimeInterval) -> Int {
    let boundedElapsed =
      clip.type == "loop" ? elapsed.truncatingRemainder(dividingBy: clip.durationSeconds) : elapsed
    let raw = Int(floor(boundedElapsed * clip.frameRate.framesPerSecond))
    return min(max(0, raw), clip.frameCount - 1)
  }

  private static func seed(from digest: String) -> UInt64 {
    UInt64(digest.prefix(16), radix: 16) ?? 0x5045_5453_4752_4150
  }
}

private struct SplitMix64 {
  private var state: UInt64

  init(seed: UInt64) { state = seed }

  mutating func next() -> UInt64 {
    state &+= 0x9e37_79b9_7f4a_7c15
    var value = state
    value = (value ^ (value >> 30)) &* 0xbf58_476d_1ce4_e5b9
    value = (value ^ (value >> 27)) &* 0x94d0_49bb_1331_11eb
    return value ^ (value >> 31)
  }

  mutating func nextUnit() -> Double {
    Double(next() >> 11) / Double(UInt64(1) << 53)
  }
}
